tests for add_tag/3 and remove_tag/3

m13m opened the pull request.

m13m committed:
add tags to a specific note

m13m committed:
remove tags from a note

m13m committed:
tests for add_tag/3 and remove_tag/3

lawik commented on lib/noted/notes.ex:
```diff
@@ -344,6 +345,53 @@ defmodule Noted.Notes do
+  def add_tag(user_id, note_id, tag_name) do
+
+    tag = Repo.get_by(Tag, name: tag_name, user_id: user_id)
+    if tag == nil do
```
If the tag exists, it may still need us to create a NotesTags entry to form the association between tag and note.

lawik replied:
We should probably check if the NotesTags row exists first. If it does we can assume the tag exists.
If the NotesTags doesn't exist we check if the tag exists.
When we are done in add_tag there should either be a Tag + NotesTags entry, or an error because something is wrong.

lawik commented on test/noted/notes_test.exs:
```diff
@@ -19,9 +19,10 @@ defmodule Noted.NotesTest do
 
-    test "list_notes/0 returns all notes" do
+    test "list_notes/1 returns all notes" do
       note = note_fixture()
```
The note_fixture should probably create the note for a specific user.

Then we can do `Notes.list_notes(note.user_id)`.

lawik commented on test/noted/notes_test.exs:
```diff
@@ -75,11 +79,56 @@ defmodule Noted.NotesTest do
+      {:ok, notes_tags} = Notes.add_tag(user.id, note.id, tag_name)
+
+      tag = Repo.get(Tag, notes_tags.tag_id)
+
```
assert that it can be loaded using get_notes_tags here.

lawik commented on test/noted/notes_test.exs:
```diff
@@ -75,11 +79,56 @@ defmodule Noted.NotesTest do
+
+      Notes.add_tag(user.id, note.id, tag_name)
+
+      assert Notes.add_tag(user.id, note.id, tag_name) == nil
```
Change add_tag return value to always return :ok | {:error, changeset}
Never nil.

Then just `assert :ok = Notes.add_tag ..`

lawik commented on test/noted/notes_test.exs:
```diff
@@ -75,11 +79,56 @@ defmodule Noted.NotesTest do
 
       tag
     end
 
```
Add a `setup do` block to add user and note to the context. See ExUnit docs for setup examples.

lawik commented on test/noted/notes_test.exs:
```diff
@@ -75,11 +79,56 @@ defmodule Noted.NotesTest do
+      Repo.one!(from n in NotesTags, where: n.note_id == ^note.id and n.tag_id == ^tag.id)
+    end
+
+    test "add_tag/3 add the tag to a note" do
```
With setup block, add , ctx for accessing context data from setup.

ctx.note, ctx.user, etc.

lawik reviewed (changes requested):
good job, some changes as discussed ✨

m13m committed:
Changes:
- update add_tag/3 is idempotent
- fixed notes_test.exs failing test
- add Makefile to create postgres docker instance.

Co-authored-by: Lars Wikman <lars@underjord.io>

lawik commented on Makefile:
```diff
@@ -0,0 +1,2 @@
+postgres:
```
I would remove this from this commit. This is one of many ways of starting Postgres and having that included indicates that it would be the recommended one and I don't particular intend to use it :)

m13m replied:
yeah make sense.

lawik commented on .gitignore:
```diff
@@ -32,3 +32,7 @@ npm-debug.log
 # this depending on your deployment strategy.
 /priv/static/
+
+.iex.exs
```
What is this from, haven't seen that before?

m13m replied:
while launching `iex` you can customize which modules to preload so saves you extra typing.

lawik commented on .gitignore:
```diff
@@ -32,3 +32,7 @@ npm-debug.log
+
+.iex.exs
+
+.env
```
This one is fine, fairly common I guess.

lawik commented on lib/noted/notes.ex:
```diff
@@ -278,6 +286,12 @@ defmodule Noted.Notes do
 
+
+
+  def get_tag_by(tag_name, user_id) do
```
The name is not very descriptive. Maybe get_tag_by_name/2 ?

The name is the important thing and you need the user to disambiguate between identically named tags of course, but I think get_tag_by_name would make sense.

lawik commented on lib/noted/notes.ex:
```diff
@@ -290,12 +304,12 @@ defmodule Noted.Notes do
 
   """
-  def create_tag!(attrs \\ %{}) do
+  def create_tag(attrs \\ %{}) do
```
Has `create_tag!` been replaced everywhere that it is used by `create_tag`? If so, that's fine. Just don't know :)

m13m replied:
it has been replaced everywhere.

lawik commented on lib/noted/notes.ex:
```diff
@@ -369,7 +385,15 @@ defmodule Noted.Notes do
+  end
+  end
+
+  defp has_notes_tags?(note_id, tag_id) do
```
Isn't this an inverse of the logic? I would expect this function to return true if it has notes tags. But it returns true if it returns an empty list.

lawik commented on lib/noted/notes.ex:
```diff
@@ -347,8 +361,10 @@ defmodule Noted.Notes do
 
   def add_tag(user_id, note_id, tag_name) do
```
I think the logic can be simplified:

- Get tag
- Check if tag exists, otherwise create tag, regardless, now we have tag or we return the error
- Check if notes_tag exists, otherwise create it, now we return :ok or the error

Also, when using Repo.transaction you should call Repo.rollback to return any error/failure.

lawik reviewed (changes requested):
Good job! Some notes to work through added.

m13m committed:
Delete Makefile

m13m committed:
refactor and function names to be explicit

lawik commented on lib/noted/notes.ex:
```diff
@@ -369,7 +385,15 @@ defmodule Noted.Notes do
+    tags
+  else
+    :ok
+  end
```
This looks like the formatter needs to be run :)

m13m replied:
ran the formatter https://github.com/lawik/noted/pull/2/commits/d0e9f1146bc594b3e659b2bf2c96499d7828854e

m13m committed:
repo rollback

lawik commented on lib/noted/notes.ex:
```diff
@@ -344,6 +359,63 @@ defmodule Noted.Notes do
+          |> Repo.insert()
+
+        error ->
+          error
```
This will not rollback the transaction. Try: `{:error, reason} -> Repo.rollback(reason)`

m13m replied:
rollback the txn changes https://github.com/lawik/noted/pull/2/commits/193e20f09eb22526394048e1a1697c159d923d5f

m13m committed:
format notes.ex
